### PR TITLE
Add dismissible compatibility notice to sidebar when editing Cart and Checkout

### LINKED ISSUE #6851: Editor: Add an alert notice about possible incompatibilities to the Cart and Checkout Blocks
## Describe the bug

The Cart and Checkout Blocks's editor sidebar will include an alert that warns about possible incompatibilities.

Figma file bA1eV1oHbHs3sEEVnXbj4h-fi-383%3A10581
Context: pdFofs-b7-p2

The alert will:
- be placed under the Name & description panel, before the Block's settings panels
- the link will point to the this [Incompatible Extensions section](https://woocommerce.com/document/cart-checkout-blocks-support-status/#section-7)
- will appear when selecting the parent Cart and Checkout Blocks as well as their inner blocks

Behaviour:
- the alert is replacing the existing [pop-up alert ](https://user-images.githubusercontent.com/1628454/182843881-7847393d-8f1e-4373-93b8-5d75e1f26bfc.png)
- the alert will appear even for users that have previously dismissed the current [pop-up alert ](https://user-images.githubusercontent.com/1628454/182843881-7847393d-8f1e-4373-93b8-5d75e1f26bfc.png)
- once it’s dismissed it wouldn’t appear again

## Expected behavior

0. As an user I create a new page and insert the Cart or the Checkout Blocks for the first time
1. I no longer see the [pop-up alert ](https://user-images.githubusercontent.com/1628454/182843881-7847393d-8f1e-4373-93b8-5d75e1f26bfc.png)
2. When I select the Block or one of it's inner blocks I notice the alert, under the name and description section in the Block's sidebar
3. I dismiss the alert and navigate away from the page
4. I come back to the page and I no longer see the alert

## Additional context

The new alert can use a dismissal mechanism similar to the one available for the current [pop-up alert](https://user-images.githubusercontent.com/1628454/182843881-7847393d-8f1e-4373-93b8-5d75e1f26bfc.png). Search for `wc-blocks_dismissed_compatibility_notices` 

The Mini Cart will still use  `wc-blocks_dismissed_compatibility_notices` so we won't remove it as part of this effort yet.

### LINKED ISSUE #6852: Editor: Update the Sidebar Feedback section for the Cart and Checkout Blocks
## Describe the bug

The Cart and Checkout Blocks's editor sidebar includes a Feedback box. We need to make it visible when people select inner blocks as well

Figma file bA1eV1oHbHs3sEEVnXbj4h-fi-383%3A10581
Context: pdFofs-b7-p2

The alert will:
- keep it's current position
- might require a text change (TBD)

## Expected behavior

0. As an user I create a new page and insert the Cart or the Checkout Blocks for the first time
2. When I select the Block or one of it's inner blocks I notice the feedback box in the Block's sidebar

### PR DESCRIPTION
**Pull request by opr**
This PR adds a notice to the inspector when editing Cart and Checkout blocks, and their inner blocks. It also adds functionality to dismiss the notice. I reused the `useCompatibilityNotice` functionality we have but used a different local storage key, thus replacing the old compatibility notice, but ensuring the new notice is visible to users who did dismiss the old one.

This PR will add the `CartCheckoutSidebarCompatibilityNotice` to the `InspectorControls` by creating a higher order component which includes it, and adding a filter to maybe include this on the `editor.BlockEdit` hook.

I also moved `BlockSettings` to a HOC to allow us to render the compatibility notice before it. This resulted in some duplicated code removal from `cart/edit.js` and `checkout/edit.tsx`.

I created the HOCs in `cart-checkout-shared`, but left the actual `CartCheckoutSidebarCompatibilityNotice` in `editor-components`.

I reimplemented the old `DefaultNotice` since @senadir is blocked by a [PR in WooCommerce Core](https://github.com/woocommerce/woocommerce/pull/34234).

Fixes #6851
Fixes #6852
Supersedes #6855
Contains changes from #6881

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

image

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Before checking this PR out, build `trunk`, clear local storage and visit the Cart/Checkout blocks. Dismiss the compatibility notices on each page. Then check out this PR.
2. Go to the editor, insert the Checkout block.
3. Select it, verify you see the notice. Click through each and every inner block and ensure the notice displays for it.
4. Dismiss the notice, click back through the inner blocks and ensure the notice is gone. Reload the page, ensure the notice is still gone.
5. Go to the editor, insert the Cart block.
6. Select it, verify you see the notice. Click through each and every inner block and ensure the notice displays for it.
7. Dismiss the notice, click back through the inner blocks and ensure the notice is gone. Reload the page, ensure the notice is still gone.
8. Clear local storage again.
9. Go back to the Cart and Checkout blocks and ensure the notice is back.
10. Create a new page. Add the Checkout Block. Select it and verify the sidebar contains the default warning:
image.
Verify the wording says **checkout**.
11. Add the Cart Block, ensure the same notice appears but ensure it says **cart**.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Changelog

> Add notice to Cart and Checkout blocks' inspector controls which links to the list of compatible plugins.

### CODE REVIEW

File: assets/js/editor-components/sidebar-compatibility-notice/use-compatibility-notice.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ export const useCompatibilityNotice = (
 	blockName: string
 ): [ boolean, () => void ] => {
 	const [ dismissedNotices, setDismissedNotices ] = useLocalStorageState(
-		`wc-blocks_dismissed_compatibility_notices`,
+		`wc-blocks_dismissed_sidebar_compatibility_notices`,
```
Comment (opr) [inline assets/js/editor-components/sidebar-compatibility-notice/use-compatibility-notice.ts:13]: Done so merchants who have dismissed the current notices will see the new one.

File: assets/js/editor-components/utils/index.js [hunk 2]
```diff
@@ -217,3 +219,29 @@ export const formatTitle = ( page, pages ) => {
...
 	return page.title.raw + ( ! isUnique ? ` - ${ page.slug }` : '' );
 };
+
+/**
+ * Gets whether the block (passed by clientId) is the Cart, Checkout, or any of their inner blocks.
+ *
+ * @param {string} clientId The clientId of the block.
+ * @return { { isCheckout: boolean, isCart: boolean } } An object with isCheckout and isCart booleans.
+ */
+export const useIsCartOrCheckoutOrInnerBlock = ( clientId ) => {
```
Comment (opr) [inline assets/js/editor-components/utils/index.js]: Maybe we need a better name for this?

Comment (senadir) [reply assets/js/editor-components/utils/index.js]: Now that this is used in a single place, do we need to have a hook?

Comment (opr) [reply assets/js/editor-components/utils/index.js]: Good point, I thought I used it in the `withBlockSettings` hoc but I didn't. Removed it now.

File: assets/js/blocks/cart-checkout-shared/block-settings/index.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		'woocommerce/add/cart-checkout-block-settings'
+	)
+) {
+	addFilter(
+		'editor.BlockEdit',
+		'woocommerce/add/cart-checkout-block-settings',
+		withBlockSettings,
+		10
+	);
+}
```
Comment (senadir) [inline assets/js/blocks/cart-checkout-shared/block-settings/index.tsx]: I don't get this abstraction, what are we trying to do here?

File: assets/js/blocks/cart-checkout-shared/default-notice/editor.scss [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+	margin: 0;
+	padding-right: 16px;
+	.components-notice__dismiss {
+		min-width: 24px;
+	}
+	svg {
+		width: 16px;
+		height: 16px;
+	}
+}
```
Comment (senadir) [inline assets/js/blocks/cart-checkout-shared/default-notice/editor.scss:11]: Currently, those styles are meant for the default cart notice (that's fine), however, can we extend them to cover the compatibility notice as well?

The goal was to match `x` sizes here
image

File: assets/js/editor-components/sidebar-compatibility-notice/index.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		</div>
+	);
+};
+
+export const CheckoutSidebarCompatibilityNotice = () => (
+	<CartCheckoutSidebarCompatibilityNotice block={ 'checkout' } />
+);
+export const CartSidebarCompatibilityNotice = () => (
+	<CartCheckoutSidebarCompatibilityNotice block={ 'cart' } />
+);
```
Comment (senadir) [inline assets/js/editor-components/sidebar-compatibility-notice/index.tsx]: Are those 2 still used?

File: assets/js/editor-components/sidebar-compatibility-notice/use-compatibility-notice.ts (hunk 1 above)
Comment (senadir) [reply assets/js/editor-components/sidebar-compatibility-notice/use-compatibility-notice.ts:13]: Dismissing the notice for Checkout doesn't dismiss it for Cart, should we unify both?

**Review by senadir — APPROVED** [review]
Thank you for working on this and cleaning up tests, much appreciated Thomas! I left some notes that can be addressed and for this to be merged after.

File: assets/js/editor-components/sidebar-compatibility-notice/index.tsx (hunk 5 above)
Comment (opr) [reply assets/js/editor-components/sidebar-compatibility-notice/index.tsx]: Good catch. No they are not.

File: assets/js/blocks/cart-checkout-shared/block-settings/index.tsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+					) }
+					checked={ hasDarkControls }
+					onChange={ () =>
+						setAttributes( {
+							hasDarkControls: ! hasDarkControls,
+						} )
+					}
+				/>
+			</PanelBody>
+			<CartCheckoutFeedbackPrompt />
```
Comment (tarhi-saad) [inline assets/js/blocks/cart-checkout-shared/block-settings/index.tsx]: @opr, previously, when selecting the `Cart` block, we could see the `feedback` notice. But, for some reason, this doesn't show up anymore!

image

EDIT 1: I believe the whole `BlockSettings` isn't added to the `Cart` block (e.g., no `Dark mode` option either).

**Review by tarhi-saad — CHANGES_REQUESTED** [review]
Thank you for your excellent work on this PR, @opr. I left a comment below!

**Comment by opr** [comment]
Latest changes:
- Remove `withBlockSettings` HOC and the filter to apply it.
- Move `BlockSettings` to `cart-checkout-shared`
- Add `BlockSettings` back to the cart and checkout edit functions
- Move `CartCheckoutFeedbackPrompt` into the `withSidebarNotices` hoc (so it appears in all inner blocks).

File: assets/js/blocks/cart-checkout-shared/block-settings/index.tsx (hunk 3 above)
Comment (opr) [reply assets/js/blocks/cart-checkout-shared/block-settings/index.tsx]: We don't have to duplicate it in cart/edit and checkout/edit. This is kind of a hangover from when we ran into the Slot/Fill bug when `CartCheckoutSidebarCompatibilityNotice` returned null.

I will move `BlockSettings` to `cart-checkout-shared` and remove this hoc. Prevents duplication without the hoc abstraction.

**Review by tarhi-saad — APPROVED** [review]

**Comment by opr** [comment]
@senadir The way I removed the need for `withAddressFields` was instead explicitly include `CartCheckoutFeedbackPrompt` within [`addressFieldControls`](https://github.com/woocommerce/woocommerce-blocks/pull/6869/files#diff-627969b1cedc2351e3e406a16c31338a6183087c251f82e491278180a2c033d9R104) and [`accountControls`](https://github.com/woocommerce/woocommerce-blocks/pull/6869/files#diff-627969b1cedc2351e3e406a16c31338a6183087c251f82e491278180a2c033d9L148)  and in the hoc ([`withSidebarNotices`](https://github.com/woocommerce/woocommerce-blocks/pull/6869/files#diff-a8b2ace3f7ee9e317dfd806c1da525369598f085d661190512e97b235103e3aeR33)) I checked if we were in a block that shows `addressFieldControls ` or `accountControls` and if so, did not render the `CartCheckoutFeedbackPrompt` a a second time.

**Review by senadir — APPROVED** [review]
This is testing fine for me

**Pull request merged**